03_letters/starfallz/main_test.go

```diff
@@ -0,0 +1,103 @@
+package main
+
+import (
+	"bytes"
+	"reflect"
+	"strconv"
+	"testing"
+)
+
+func TestMainFunction(t *testing.T) {
+	t.Run("Test main to return string values of letters(aba) with proper formatting", func(t *testing.T) {
+		var buf bytes.Buffer
+		out = &buf
+
+		main()
+
+		expected := strconv.Quote(`a:2
+b:1`)
+		actual := strconv.Quote(buf.String())
+
+		if expected != actual {
+			t.Errorf("Unexpected output, expected: %s, actual: %s", expected, actual)
+		}
+	})
+}
+
+func TestLettersFunction(t *testing.T) {
+	testCases := []struct {
+		description string
+		input       string
+		expected    map[rune]int
+	}{
+		{"Test letters to return sorted map of letters based on rune # with count", "abaabazzzzzz",
+			map[rune]int{97: 4, 98: 2, 122: 6}},
+		{"Test letters to handle uppercase letters", "aAAaA", map[rune]int{65: 3, 97: 2}},
+		{"Test letters to handle non-alphabetical characters", "    !!!*", map[rune]int{32: 4, 33: 3, 42: 1}},
+		{"Test letters to return empty map on empty string input", "", map[rune]int{}},
+	}
+
+	for _, testCase := range testCases {
+		input := testCase.input
+		expected := testCase.expected
+		t.Run(testCase.description, func(t *testing.T) {
+			actual := letters(input)
+
+			if !reflect.DeepEqual(actual, expected) {
+				t.Errorf("Unexpected output, expected: %d, actual: %d", expected, actual)
+			}
+		})
+	}
+}
+
+func TestSortLettersFunction(t *testing.T) {
+	testCases := []struct {
+		description string
+		input       map[rune]int
+		expected    []string
+	}{
+		{"Test sort letters using letters() as input and return count", letters("aba"), []string{"a:2", "b:1"}},
+		{"Test sort letters with uppercase using letters() as input and return count", letters("Aba"),
+			[]string{"A:1", "a:1", "b:1"}},
+		{"Test sort letters with uppercase and return count", map[rune]int{97: 2, 65: 3}, []string{"A:3", "a:2"}},
```

Comment:
i would prefer to keep rune as rune (i.e. `'a':2 `instead of `97:2`

Reply:
Cool no worries will change them. It's easier to read too. Thanks

Reply:
Would make it much easier to read the weird Unicode too. By definition Go source code is UTF-8 encoded, so you shouldn't be afraid of putting weird characters in, every tool used should either support it or be abandoned.